Update: allow shadowed references in no-alert (fixes #1105)

lo1tuma opened the pull request.

ilyavolodin commented:
This isn't going to cover shadowing from the upper scope, right?
```
var alert = function() {};
function test() {
    alert();
}
````
You either have to loop through all scopes all the way to the global, or subscribe to `VariableDeclaration` and `AssignmentExpression` and check if `id.name`/`left.name` is alert.

michaelficarra commented on lib/rules/no-alert.js:
```diff
@@ -8,44 +10,123 @@
-            } else if (node.callee.type === "MemberExpression" && node.callee.property.type === "Identifier") {
-
-                result = matchProhibited(node.callee.property.name);
+            } else if (callee.type === "MemberExpression" && callee.object.name === "window") {
```
`window` is not necessarily the global object. Not only may it be a shadowed reference, but that's just not a standard property of the global object.

lo1tuma replied:
Good point. Should we additionally check if `window` is a implict global and not shadowed?

michaelficarra replied:
How would you know that it exists on the global object statically?

ilyavolodin replied:
We add it to the global scope when browser environment is enabled. (I think)

lo1tuma replied:
ESLint adds this to the list of variables of the global scope if you enable the browser environment.

michaelficarra replied:
You don't know that this is a reference to the global object, though. I could have a global called `window` in my node application, and it could be a GUI window object, not a reference to the global object.

lo1tuma commented:
Good catch.

A loop is not necessary. I have to check only to global scope for references that don’t have a resovled variable. I've updated the PR and added your example to the tests.

michaelficarra commented:
I'd like to see `this.alert` at the top level in sloppy mode caught by this rule.

lo1tuma commented:
@michaelficarra actually this was not requested in #1105, but I see your point. Any idea how to achieve this? If this is limited to top level it shouldn’t be too hard IMHO.

lo1tuma commented:
I've updated the PR and addressed the feedback from @michaelficarra.

ilyavolodin commented on tests/lib/rules/no-alert.js:
```diff
@@ -22,14 +24,87 @@ eslintTester.addRuleTest("lib/rules/no-alert", {
+        "function foo() { var alert = function() {}; function test() { alert(); } }",
+        "function confirm() {} confirm();",
+        "function prompt() {} prompt();",
+        "window[alert]();",
```
Wait, why it this valid? Sorry, I might've missed something.

lo1tuma replied:
This is a computed `MemberExpression`.

`window[alert] !== window["alert"]`.

ilyavolodin replied:
Got it. So it's just a sanity check.

lo1tuma replied:
Correct. It also covers the else branch in `getPropertyName` where the `MemberExpression` is computed but the property is not a `Literal`.

michaelficarra commented:
@lo1tuma

> actually this was not requested in #1105

Neither was `window.alert`, but that is in this implementation. If we have that, we should support things like `window.window.window.alert` and `this.alert` at the top level. But I don't think we should catch `window.alert` because it relies on a convention that we cannot trust.

lo1tuma commented:
> Neither was window.alert, but that is in this implementation.

I haven't implemented this feature. This is in ESLint since version 0.0.5 (see: 967fcd570e60544f76cbd2b2a45a2462922d6717). If we remove this it would be a breaking change and probably out of scope of this PR.

michaelficarra commented:
Oh, sorry. I didn't notice that.

ilyavolodin commented:
I think it's good to merge. @michaelficarra I think if you feel strongly that those edge cases should be covered, we can probably do them under a separate issue.

michaelficarra commented:
Agreed, let's keep it out of this PR.
